Step2 로또(자동)

goodbyeyo opened the pull request:
안녕하세요 

자동 로또 리뷰 부탁드립니다.

goodbyeyo committed:
docs: 자동 로또 미션 요구사항 정리

goodbyeyo committed:
feat: Lotto 도메인 테스트 및 구현

goodbyeyo committed:
feat: 로또 번호, 티켓 발급 기능 구현 및 테스트

goodbyeyo committed:
refactor: Test 코드 내 도메인 로직 Class 분리 리팩토링

goodbyeyo committed:
feat: 당첨 로또 기능 구현

vsh123 commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -0,0 +1,19 @@
+object InputView {
+    fun getUserAmount(): Int {
+        println("구입 금액을 입력해 주세요.")
+        val amount: String = readln()
```
kotlin의 타입추론을 이용해보면 어떨까요?

vsh123 commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -0,0 +1,19 @@
+    fun getUserAmount(): Int {
+        println("구입 금액을 입력해 주세요.")
+        val amount: String = readln()
+        return amount.toInt()
```
[toIntOrNull](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/to-int-or-null.html)을 활용해서, 숫자가 아닌경우에 대한 처리를 해보면 어떨까요?

vsh123 commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -0,0 +1,20 @@
+import lotto.domain.LottoTickets
+import lotto.domain.WinningLotto
+
+class LottoController {
```
[object](https://kotlinlang.org/docs/object-declarations.html)를 활용해봐도 좋을 것 같아요!

vsh123 commented on src/main/kotlin/lotto/domain/LottoTickets.kt:
```diff
@@ -0,0 +1,42 @@
+    companion object {
+        const val LOTTO_TICKET_PRICE = 1000
+
+        fun purchase(amount: Int): LottoTickets {
```
amout나 quantity 의 최소값 기준은 없을까요??

vsh123 commented on src/main/kotlin/lotto/domain/LottoTickets.kt:
```diff
@@ -0,0 +1,42 @@
+
+        fun purchase(amount: Int): LottoTickets {
+            val quantity = amount / LOTTO_TICKET_PRICE
+            val lottoTickets = ArrayList<LottoTicket>()
+            repeat(quantity) {
+                lottoTickets.add(LottoTicket.generateLottoNumber())
+            }
```
```suggestion
            val lottoTickets = List(quantity) { LottoTicket.generateLottoNumber()}
```

위 처럼 작성이 가능해 보여요! :)

vsh123 commented on src/main/kotlin/lotto/domain/LottoTickets.kt:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    companion object {
+        const val LOTTO_TICKET_PRICE = 1000
```
private 선언이 가능해보이네요!

vsh123 commented on src/main/kotlin/lotto/domain/LottoTickets.kt:
```diff
@@ -0,0 +1,42 @@
+        _tickets = lottoTickets
+    }
+
+    fun calculateLottoRank(winningLotto: WinningLotto): LottoResults {
```
해당 테스트가 Controller에 있지만, Tickets에도 단위테스트를 만들어 검증해보면 어떨까요?

vsh123 commented on src/main/kotlin/lotto/domain/LottoTickets.kt:
```diff
@@ -0,0 +1,42 @@
+package lotto.domain
+
+class LottoTickets(lottoTickets: List<LottoTicket>) : Collection<LottoTicket> by lottoTickets {
```
위임을 활용하셨군요!! 👍  다만 lottoTickets은 생성자의 인자일뿐 프로퍼티가 아닌데요!

프로퍼티를 기반으로 위임을 진행해야 하지 않을까요?

vsh123 commented on src/test/kotlin/lotto/domain/LottoTicketTest.kt:
```diff
@@ -0,0 +1,54 @@
+    @Test
+    fun `로또 티켓을 생성하면 6개의 번호가 들어있다`() {
+        val lottoTicket = LottoTicket.generateLottoNumber()
+        lottoTicket.lottoNumbers.size shouldBe 6
```
shouldHaveSize를 이용해보면 좋을 것 같아요!

vsh123 commented on src/test/kotlin/lotto/domain/LottoTicketsTest.kt:
```diff
@@ -0,0 +1,23 @@
+    fun `로또 티켓을 구입 금액에 맞게 발급한다`() {
+        val amount = 5_000
+        val lottoTickets = LottoTickets.purchase(amount)
+        lottoTickets.tickets.size shouldBe 5
```
위임을 사용했다면 아래와 같이 사용이 가능해보여요 :) 
```suggestion
        lottoTickets shouldHaveSize 5
```

vsh123 reviewed (changes requested):
안녕하세요!
step2 미션 작성 잘해주셨네요 💯 

몇가지 코멘트를 남겼으니 확인부탁드립니다!

goodbyeyo committed:
feat: 로또 컨트롤러, 입출력 뷰 구현

goodbyeyo committed:
refactor: 불필요한 메서드 제거

goodbyeyo committed:
refactor: 피드백 반영
1. 타입추론 이용으로 자료형 제거
2. 구입금액 숫자가 아닌 경우 처리

goodbyeyo committed:
refactor: 피드백 반영, 최소 구입 금액 체크

goodbyeyo committed:
refactor: 피드백 반영
1. repeat 연산자를 List 사용하도록 변경

goodbyeyo committed:
refactor: 피드백 반영, LottoTickets.calculateLottoRank 테스트 추가

goodbyeyo committed:
refactor: 피드백 반영
1. LottoTickets 프로퍼티를 기반으로 위임하도록 변경
2. LottoTicket 위임을 유지하면서 생성자 값 명시적인 변환

vsh123 commented on src/main/kotlin/lotto/domain/WinningLotto.kt:
```diff
@@ -0,0 +1,19 @@
+
+import lotto.domain.LottoTicket.Companion.LOTTO_TICKET_SIZE
+
+class WinningLotto(numbers: List<Int>) {
```
LottoTicket과 차이점이 있을까요? 두개의 클래스를 분리하신 이유가 궁금합니다!

vsh123 commented on src/main/kotlin/lotto/domain/LottoTicket.kt:
```diff
@@ -0,0 +1,35 @@
+import lotto.domain.LottoNumber.Companion.LOTTO_MIN_NUMBER
+
+class LottoTicket(numbers: List<Int>) : Collection<Int> by numbers.toSet() {
+    private val _lottoNumbers: Set<Int> = numbers.toSortedSet()
+    val lottoNumbers: Set<Int>
+        get() = _lottoNumbers.toSet()
```
해당 로직에 backing properties를 적용하신 이유가 있을까요?

vsh123 reviewed (changes requested):
안녕하세요! 
리뷰가 늦어서 죄송합니다!!

코멘트 남겼으니 확인부탁드립니다 :)

goodbyeyo committed:
feat: LottoNumber 정적팩토리메서드 추가

goodbyeyo committed:
refactor: 피드백 반영
1. WinningLotto 제거하고 담당하던 책임 LottoTicket 으로 이동
2. LottoTicket 프로퍼티 변경 : List<Int> -> Set<LottoNumber>
3. LottoTicket 프로퍼티 기반으로 위임하도록 변경, 부생성자 추가
4. LottoTicket 정적팩토리메서드 추가

vsh123 commented on src/main/kotlin/lotto/domain/LottoTicket.kt:
```diff
@@ -0,0 +1,52 @@
+        return numbers.intersect(winningLotto).size
+    }
+
+    override fun equals(other: Any?): Boolean {
```
equlas와 hashcode를 직접 구현했다는건, data class로의 선언을 고민해봐도 좋을 것 같아요!!

고민한번 부탁드릴게요~ :)

vsh123 reviewed (approved):
안녕하세요!
step2 작성하시느라 고생많으셨습니다 :) 

코멘트 하나 남겼으니 다음 스텝에서 함께 고민부탁드릴게요!

이번 스텝은 마무리하도록 하겠습니다!
